### DISCUSSION TITLE
Can MOOSE Handle Salt Precipitation and Its Induced Porosity and Permeability Changes?

### DISCUSSION THREAD

**Discussion by robo-boluo** [idaholab/moose#27832] [Q&A Modules: Porous Flow]
Dear MOOSE Team,

I am writing to inquire about the capabilities of MOOSE regarding NaCl precipitation during CO₂ injection into a brine aquifer. Specifically, I am interested in understanding whether the current version of MOOSE can model the effects of salt precipitation on porosity and permeability changes.

If this functionality is not currently available, could you provide guidance on how to implement this feature?

Thank you for your assistance.

**Comment by GiudGiud** [comment]
Hello

Porosity and permeability are both material properties which can be set dynamically during the simulation as a function of NaCl precipitate concentration. 
So yes, you could model this. Implementing new material properties should be (relatively) straightforward after going through the tutorial

Guillaume

**Comment by robo-boluo** [comment]
Hi GiudGiud, @GiudGiud  @cpgr

I want to express my gratitude for your valuable advice. In response to it, I created new source files 'PorousFlowPorosity_nacl.h' and 'PorousFlowPorosity_nacl.C' ([link](https://drive.google.com/drive/folders/11RLH6zaP1aGQkqJNdjX2Nctd1dWTgzAb?usp=drive_link))to incorporate salt precipitation.

However, upon implementing PorousFlowPorosity_nacl in the [Materials] block, I encountered an issue during testing. It appears that '(PorousFlow_density_qp, dPorousFlow_density_qp_dvar, PorousFlow_xnacl_qp, dPorousFlow_xnacl_qp_dvar, PorousFlow_density_nodal, dPorousFlow_density_nodal_dvar, PorousFlow_xnacl_nodal, dPorousFlow_xnacl_nodal_dvar)' are not defined on the specified blocks (caps and aquifer).

I called the PorousFlowPorosity_nacl using the following format:
```
   [porosity]
    type = PorousFlowPorosity_nacl
    block=aquifer
    fluid = true
    mechanical = true
    thermal = true
    salt_precip = true
    porosity_zero = 0.2
    reference_temperature = 105
    reference_porepressure = 15.5E6
    thermal_expansion_coeff = 15E-6 # volumetric
    solid_bulk = 8E9 # unimportant since biot = 1
    xnacl_L = 0.30
    rho_nacl = 2165.0
    r_m = 1.0e-5
    xnacl=xnacl
    density=density_h2o
  []
```
Could you please provide suggestions on how to address this issue at your convenience?

Thank you very much!

**Reply by GiudGiud** [reply]
hello

use Debug/show_material_props to see which properties are being defined, and which are missing based on the current error you are getting. 

it might be that they are all being defined, but the names are not quite what is expected

**Reply by robo-boluo** [reply]
Hi GiudGiud,

Thank you for your prompt response.

The codes currently are not executable, so the [debug] does not show any info.

**Reply by GiudGiud** [reply]
Comment out the material giving the error to try to get far enough into the setup that
```
[Debug]
  show_material_props = true
[]
```
shows information

**Reply by robo-boluo** [reply]
Hello GiudGiud @GiudGiud,

I am grateful for your kind response to my previous question. 

I'm a bit confused about how to implement the variable 'xnacl' to my new 'PorousFlowPorosity_nacl' application. I've tried two methods to implement 'xnacl', but I'm still encountering issues. Could you please provide some guidance on this matter?



1) Take 'xnacl' as a  MaterialProperty, similar to 'pf' fluid pressure 
.h file
```
  const MaterialProperty<Real> * const _xnacl;
  const MaterialProperty<std::vector<Real>> * const _dxnacl_dvar;
```
and .C file
```
    _xnacl(_salt_precip ? (_nodal_material
                      ? &getMaterialProperty<Real>("PorousFlow_mass_frac_nodal")
                      : &getMaterialProperty<Real>("PorousFlow_mass_frac_qp"))
               : nullptr),
    _dxnacl_dvar(_salt_precip ? (_nodal_material ? &getMaterialProperty<std::vector<Real>>(
                                              "dPorousFlow_mass_frac_nodal_dvar")
                                        : &getMaterialProperty<std::vector<Real>>(
                                              "dPorousFlow_mass_frac_qp_dvar"))
                     : nullptr),

```
2) Take 'xnacl' as a  MaterialProperty, similar to '_saturation' fluid saturation 
.h file
```
  const MaterialProperty<std::vector<Real>> * const _xnacl;
  const MaterialProperty<std::vector<std::vector<Real>>> * const _dxnacl_dvar;
```
and .C file
```
     _xnacl(_salt_precip 
                    ? (_nodal_material
                           ? &getMaterialProperty<std::vector<Real>>("PorousFlow_mass_frac_nodal")
                           : &getMaterialProperty<std::vector<Real>>("PorousFlow_mass_frac_qp"))
                    : nullptr),
    _dxnacl_dvar(_salt_precip 
                          ? (_nodal_material ? &getMaterialProperty<std::vector<std::vector<Real>>>(
                                                   "dPorousFlow_mass_frac_nodal_dvar")
                                             : &getMaterialProperty<std::vector<std::vector<Real>>>(
                                                   "dPorousFlow_mass_frac_qp_dvar"))
                          : nullptr),
```

**Reply by GiudGiud** [reply]
Hello

Are you planning to solve for x_nacl ? Or can it be computed directly from the other quantities being solved for?

So you are asking if it should be a Real (one float value per quadrature point) or a std::vector<Real> (a vector of values on every quadrature point) ?
Is there any reason it should be a vector? Does every phase have a different x_nacl ?

**Reply by robo-boluo** [reply]
Hello @GiudGiud

NaCl only exists in the aqueous phase. In my developed 'PorousFlowPorosity_nacl' app, I plan to use the mass fraction 'x_nacl' as a variable. However, I am unsure how to integrate 'x_nacl' into my 'PorousFlowPorosity_nacl' app. Typically, 'x_nacl' is defined in [FluidProperties], [UserObjects], and [Materials]. 


```
[FluidProperties]
  [co2]
    type = CO2FluidProperties
  []
  [brine]
    type = BrineFluidProperties
  []
[]

[UserObjects]
  [dictator]
    type = PorousFlowDictator
    porous_flow_vars = 'pgas zi xnacl temperature disp_r'
    number_fluid_phases = 2
    number_fluid_components = 3
  []
  [pc]
    type = PorousFlowCapillaryPressureConst
    pc = 0
  []
   [fs]
    type = PorousFlowBrineCO2
    brine_fp = brine
    co2_fp = co2
    capillary_pressure = pc
  []
[]

[Materials]
  [brineco2]
    type = PorousFlowFluidState
    gas_porepressure = pgas
    z = zi
    temperature = temperature
    temperature_unit = Celsius
    xnacl = xnacl
    capillary_pressure = pc
    fluid_state = fs
  []
[]

```

**Reply by GiudGiud** [reply]
It is being specified as a variable to the PorousFlowDictator ?
If it s a variable, then the question is do you solve for it, or can you compute it from the other quantities?
This determines if it should be a Variable or an AuxVariable

**Reply by robo-boluo** [reply]
It is a variable defined in [Variables] and is used to compute 'MassTimeDerivative', 'AdvectiveFlux', and 'DispersiveFlux' in [Kernels], similar to how temperature and pressure variables are used. You can access the input file via this [link](https://drive.google.com/drive/folders/11RLH6zaP1aGQkqJNdjX2Nctd1dWTgzAb?usp=drive_link) .

The application of 'xnacl' follows the examples provided in the MOOSE folder. The current issue is how to incorporate 'xnacl' into my developed 'PorousFlowPorosity_nacl.h' and 'PorousFlowPorosity_nacl.C' files.

```
Kernels]
  [mass]
    type = PorousFlowMassTimeDerivative
    fluid_component = 2
    variable = xnacl
  []
  [flux]
    type = PorousFlowAdvectiveFlux
    fluid_component = 2
    variable = xnacl
    use_displaced_mesh = false
  []
  [diff]
    type = PorousFlowDispersiveFlux
    fluid_component = 2
    variable = xnacl
    disp_long = '0 0'
    disp_trans = '0 0'
  []
[]

```

**Reply by GiudGiud** [reply]
Why are you retrieving it as a material property then?

I understand for the derivatives.

If it s a variable then Real is the right type

**Reply by robo-boluo** [reply]
I attempted to retrieve 'xnacl' as a material property, but it did not work. Could you please guide me on how to retrieve 'xnacl' for my 'PorousFlowPorosity_nacl' application? 
Thank you.

**Reply by GiudGiud** [reply]
Retrieve it as a variable since it's a variable. 

addCoupledVar in the valudParams 
And coupledValue in the constructor. 

There are examples, for example the coupledForce kernel

**Reply by robo-boluo** [reply]
Hello Guillaume Giudicelli, @GiudGiud 

Thank you for your kind advice.

I used 'xnacl' as a variable following the 'coupledForce' kernel format. In the input.i file, I called 'PorousFlowPorosity_nacl' in the [Materials] block. However, I did not observe any porosity reduction in the simulation results. In addition, 'xnacl' exceeded the assigned maximum NaCl solubility, which is definitely incorrect.

In my 'PorousFlowPorosity_nacl.C' file, halite starts to precipitate once 'xnacl' exceeds 'xnacl_L', where 'xnacl_L' is the maximum solubility in brine. The halite precipitation rate is described by Equation 6 in Grimm et al. (2020, [doi ](https://doi.org/10.1029/2020WR027213)). In my implementation of 'PorousFlowPorosity_nacl', the new variable is 'xnacl', and the new constants are '_xnacl_L' (maximum mass fraction solubility of NaCl), '_rho_nacl' (halite density), '_r_m' (precipitation rate), and the boolean variable 'salt_precip'.

Could you please take a quick look at the codes ([link](https://drive.google.com/drive/folders/11RLH6zaP1aGQkqJNdjX2Nctd1dWTgzAb?usp=drive_link))at your convenience?

Your time and help are greatly appreciated.

**Reply by GiudGiud** [reply]
Hello

I just requested access to that link.
I am curious how you implemented the bound value on xnacl

Guillaume

**Reply by robo-boluo** [reply]
Thank you for your prompt response. I have just granted access. Please let me know if you encounter any issues accessing it.

**Reply by GiudGiud** [reply]
I just took a look. I see the porosity is modified depending on where x_nacl sits compared to xnacl_L.
However I dont see a kernel or a constraint preventing xnacl from going past xnacl_L ?

```
  [mass2]
    type = PorousFlowMassTimeDerivative
    fluid_component = 2
    variable = xnacl
  []
  [flux2]
    type = PorousFlowAdvectiveFlux
    fluid_component = 2
    variable = xnacl
    use_displaced_mesh = false
  []
  [diff2]
    type = PorousFlowDispersiveFlux
    fluid_component = 2
    variable = xnacl
    disp_long = '0 0'
    disp_trans = '0 0'
  []
  [vol_strain_rate2]
    type = PorousFlowMassVolumetricExpansion
    fluid_component = 2
    variable = xnacl
  []
```

**Reply by robo-boluo** [reply]
That is correct. I have questions about implementing an algorithm to prevent 'xnacl' from exceeding 'xnal_L'. The only modification in my input.i file is the porosity in the [Materials] block. Do you think the NaCl mass fraction constraint should be defined in the source code (.h and .C files) or in the input.i file?

**Reply by GiudGiud** [reply]
Constraints can be defined in the input file.
I think this would be a good use for the Bounds system. See this object to add to your simulation:
https://mooseframework.inl.gov/source/bounds/ConstantBounds.html

You also enforce it by writing a custom force kernel which would gradually penalize the residual when xnacl goes out of bounds.

**Reply by robo-boluo** [reply]
Thanks, I will update to you later.

**Comment by robo-boluo** [comment]
Hi Guillaume Giudicelli, @GiudGiud 

This is a quick update of the addition of ConstantBounds in the input file ([link](https://drive.google.com/drive/folders/11RLH6zaP1aGQkqJNdjX2Nctd1dWTgzAb?usp=sharing)).  Two simulation results are observed.

(1) Regarding the mass fraction shuffle

At Timestep 31 in Fig.1, the injection inlet reaches the maximum mass fraction of (xnacl_L= 0.3). Subsequently, there is a mass fraction shuffle in the next timestep (as seen in Fig.2). Following this, the NaCl propagates further into the formation, maintaining a maximum mass concentration of 0.3 (as depicted in Fig.3 and Fig.4). The mass fraction shuffle refers to the redistribution or rearrangement of NaCl within the system. How to intepret the mass fraction shuffle?

![image](https://github.com/idaholab/moose/assets/59897601/776d94ed-85e8-43d0-b88c-22fe1ddf6849)
Fig.1

![image](https://github.com/idaholab/moose/assets/59897601/bc353ace-9055-4c33-ac12-0bc7ba290ec7)
Fig.2

![image](https://github.com/idaholab/moose/assets/59897601/1766f5b8-fa33-4e1b-9dae-5e1f7babd4f8)
Fig.3

![image](https://github.com/idaholab/moose/assets/59897601/520baa64-a55b-4125-bac0-afe96cb443cb)
Fig.4

(2) Regarding the lack of porosity changes in the reservoir layer

Despite reaching its limit (xnacl_L), no porosity reduction is observed in the reservoir layer (see Fig.5)

![image](https://github.com/idaholab/moose/assets/59897601/2adc34b4-b1f9-424a-8018-77d1c856edcf)
Fig.5

**Reply by GiudGiud** [reply]
these results do not look converged
```
  nl_abs_tol = 1e-7
  nl_rel_tol = 1e-5
```
is probably too loose. 

(2) Is this the output from your material? Is it possible that the logic you put in the material is not correct?
which part in the expression changes based on xna_cl being too big?

**Reply by robo-boluo** [reply]
Thanks for the prompt reply.

(1). Will add more strict limit.


(2) In the PorousFlowPorosity_nacl.C file, line 207 and line 229 when compute the atNegInfinityQp and datNegInfinityQp respectively. 

```

Real
PorousFlowPorosity_nacl::atNegInfinityQp() const
{
  Real result = _biot; 
  if (_salt_precip)
  { 
    const Real delta_xnacl = ((_xnacl)[_qp] - _xnacl_L);
    if (delta_xnacl > 0)
    {
      _halite_conc += _dt * _r_m * (*_porosity_old)[_qp] * (*_saturation)[_qp][_aq_ph] * (*_density)[_qp][_aq_ph] * delta_xnacl / _rho_nacl;
      result -= _halite_conc;
    }
  }
  return result;
}




Real
PorousFlowPorosity_nacl::datNegInfinityQp(unsigned pvar) const
{
  Real result = 0.0;
  if (_salt_precip && (_t_step >= 1 || _app.isRestarting()))
  { 
    const Real delta_xnacl = ((_xnacl)[_qp] - _xnacl_L);
    if (delta_xnacl > 0)
    {
      result -= _dt * (*_porosity_old)[_qp] *
                ( _r_m * (*_density)[_qp][_aq_ph] * delta_xnacl / _rho_nacl * (*_dsaturation_dvar)[_qp][_aq_ph][pvar] 
                + (*_saturation)[_qp][_aq_ph] * _r_m / _rho_nacl * ((*_ddensity_dvar)[_qp][_aq_ph][pvar] * delta_xnacl + (_grad_xnacl)[_qp] * (*_density)[_qp][_aq_ph]) );
    }
  }
  return result;
}

```


Porosity is an an AuxVariables and defined in the [AuxKernels] block (line 193) in the input.i file:

```
[AuxKernels] 
  [porosity]
    type = PorousFlowPropertyAux
    variable = porosity
    property = porosity
    execute_on = timestep_end
  []
[]
```
And the porosity is also used In the [Materials] block for aquifer and caprocks:

```

  [porosity_aquifer]
    type = PorousFlowPorosity_nacl
    block=aquifer
    fluid = true
    mechanical = true
    thermal = true
    salt_precip = true
    porosity_zero = 0.2
    reference_temperature = 105
    reference_porepressure = 15.5E6
    thermal_expansion_coeff = 15E-6 # volumetric
    solid_bulk = 8E9 # unimportant since biot = 1
    xnacl_L = 0.30
    rho_nacl = 2165.0
    r_m = 1.0e-2
    xnacl=xnacl
  []

  [porosity_caps]
    type = PorousFlowPorosity_nacl
    block=caps
    fluid = true
    mechanical = true
    thermal = true
    salt_precip = true
    porosity_zero = 0.1
    reference_temperature = 105
    reference_porepressure = 15.5E6
    thermal_expansion_coeff = 15E-6 # volumetric
    solid_bulk = 8E9 # unimportant since biot = 1
    xnacl_L = 0.30
    rho_nacl = 2165.0
    r_m = 1.0e-2
    xnacl=xnacl
  []

```

**Reply by GiudGiud** [reply]
> (2) In the PorousFlowPorosity_nacl.C file, line 207 and line 229 when compute the atNegInfinityQp and datNegInfinityQp respectively.

But the bounds will prevent it from going over the limit? So it will never affect these lines

**Reply by robo-boluo** [reply]
Yes, this is the tricky part.

Firstly, we should use the ConstantBounds to constrain xnacl ≤ xnacl_L because of the mechanism of maximum solubility. Secondly, if we limit the value of xnacl, no halite will be produced.

**Reply by GiudGiud** [reply]
> Firstly, we should use the ConstantBounds to constrain xnacl ≤ xnacl_L because of the mechanism of maximum solubility. Secondly, if we limit the value of xnacl, no halite will be produced.

exactly. So I would say:
- dont use bounds
- use a physical mechanism like a "Rate" term (constant? or propoertional to (x_nacl-limit), whichever is most physical) in your equations which is only active after xnacl > xnacl_L and works to reduce x_nacl

**Reply by robo-boluo** [reply]
I agree with your suggestion. However, I still have one question. After the halite precipitation, I can update the new xnacl_new in PorousFlowPorosity_nacl.C using the formula:
```
𝑥𝑛𝑎𝑐𝑙_𝑛𝑒𝑤=𝑥𝑛𝑎𝑐𝑙_𝑜𝑙𝑑−(𝑥𝑛𝑎𝑐𝑙_𝑜𝑙𝑑−𝑥𝑛𝑎𝑐𝑙_𝐿)×rate
```
My question is, how do I return the xnacl_new value to the model for the next time step calculation?

**Reply by GiudGiud** [reply]
no you should not do this in the material property.
You should code this mechanism in the equations, and the material property just reads the current value of xnacl

**Reply by robo-boluo** [reply]
Hello Guillaume Giudicelli, @GiudGiud ,

This is quick update after coding the xnacl update equation (𝑥𝑛𝑎𝑐𝑙_𝑛𝑒𝑤=𝑥𝑛𝑎𝑐𝑙_𝐿=0.3) for [Kernals]:


Code for XNaclUpdate.h:

```
#pragma once

#include "Kernel.h"

class XNaclUpdate : public Kernel
{
public:
  static InputParameters validParams();
  XNaclUpdate(const InputParameters & parameters);

protected:
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
private:
  const VariableValue & _xnacl;
  const Real _xnacl_L;
};
```


Code for XNaclUpdate.C:

```
#include "XNaclUpdate.h"
#include "MooseVariable.h"

registerMooseObject("PorousFlowApp", XNaclUpdate);

InputParameters
XNaclUpdate::validParams()
{
  InputParameters params = Kernel::validParams();
  params.addCoupledVar("xnacl", "The salt mass fraction in the brine (kg/kg)");
  params.addParam<Real>("xnacl_L", "nacl solubility limit at specific temp and pressure");
  return params;
}

XNaclUpdate::XNaclUpdate(const InputParameters & parameters) :
  Kernel(parameters),
  _xnacl(coupledValue("xnacl")),
  _xnacl_L(getParam<Real>("xnacl_L"))
{
}

Real
XNaclUpdate::computeQpResidual()
{
  // Compute the residual contribution
  Real residual = 0.0;
  if (_xnacl[_qp] > _xnacl_L)
  {
    residual = (_xnacl_L - _xnacl[_qp]) * _test[_i][_qp];
  }
  return residual;
}

Real
XNaclUpdate::computeQpJacobian()
{
  // Compute the Jacobian contribution
  Real jacobian = 0.0;
  if (_xnacl[_qp] > _xnacl_L)
  {
    jacobian = -_phi[_j][_qp] * _test[_i][_qp];
  }
  return jacobian;
}
```

Resutls:

(1) There are on changes for porosity

![image](https://github.com/idaholab/moose/assets/59897601/ad2d129f-e414-445d-a890-05b61bc3e524)

(2) xnacl can exceed the solubility limit xnacl_L

![image](https://github.com/idaholab/moose/assets/59897601/efed386b-53d5-4902-87a9-26328d056cef)

(3) Not converge after xnacl>xnacl_L, as timestep <e-4


Do you think what else can we do to further improve the simulation results? 

Appreciate your time and guidance.

**Reply by GiudGiud** [reply]
1) that's expected, porosity is averaged and xnacl is on average less than 0.3 still
2) it can exceeed because the rate of removal is 1 * (xnacl- limit). Maybe there should be a larger rate? especially if physically it can't exceed the limit. Add a coefficient
3) This is more troubling. Maybe use `-pc_type lu` to see if you can get it to converge

**Reply by robo-boluo** [reply]
Hi Guillaume Giudicelli, @GiudGiud 

A coefficient of 2 has been added to the 'XNaclUpdate' kernel, and the '-pc_type lu' has already been included in the [Preconditioning] block of the previous input file. However, there is no difference compared to the previous case with the coefficient =1; the three issues mentioned above still persist.

Regarding the comment that an average xnacl < 0.3 causes no porosity change, I removed the XNaclUpdate  Kernel and allowed the xnacl to increase without restriction. However, this example still yields non-changeable reservoir porosity. Therefore, I suspect that the implementation of 'PorousFlowPorosity_nacl' is incorrect.

![image](https://github.com/idaholab/moose/assets/59897601/8bf9803e-7275-407d-8fc6-8e800a8dee61)

Figure 1.  The xnacl profile without maximum solubility restriction still yields non-changeable reservoir porosity

**Reply by GiudGiud** [reply]
What steps have you taken to verify the implementation of PorousFlowPorosity_nacl ?
You dont need to run a full simulation to check it by the way.

You can create a dummy input with a Xnacl auxiliary variable. You can see this variable to an arbirtary value using a ParsedAux auxkernel

**Reply by robo-boluo** [reply]
Hi Guillaume Giudicelli @GiudGiud ,

Thank you for your continuous support.

**1. PorousFlowPorosity_nacl Verification**
I developed the 'PorousFlowPorosity_nacl' model based on the existing 'PorousFlowPorosity' model in the MOOSE folder (/moose/modules/porous_flow/). I implemented the salt precipitation effect by following the format of the chemical part in the PorousFlowPorosity.C file, as both mechanisms consider solid volume change. Consequently, I developed the code for salt precipitation under the 'atNegInfinityQp', 'datNegInfinityQp', 'atZeroQp', and 'datZeroQp' sections.
The basic mathematical equation to describe the salt precipitation is:

```
𝜙𝑛+1=𝜙𝑛−𝜙𝑛⋅𝑆𝑤⋅𝑑𝑡⋅𝜌𝑤(𝑥𝑛𝑎𝑐𝑙-𝑥𝑛𝑎𝑐𝑙_𝐿)⋅𝑟𝑚⋅𝜌𝑚
```

where 𝑆𝑤​ and 𝜌𝑤 are the aqueous saturation and density, 𝑟𝑚 is the precipitation rate, and 𝜌𝑚 is the halite density.

**2. Dummy Test**

I tested the following dummy input file, and the code is executable. 


```
[Mesh]
  type = GeneratedMesh 
  dim = 2
  nx = 10
  ny = 10
  xmin = 0
  xmax = 1
  ymin = 0
  ymax = 1
[]

[GlobalParams]
  PorousFlowDictator = dictator
[]

[Variables]
  [pwater]
    initial_condition = 1e6
  []
  [xnacl]
    initial_condition = 0.1
  []
  [zi]
    initial_condition = 0
  []
[]

[AuxVariables]
  [aux_xnacl]
    order = CONSTANT
    family = MONOMIAL
  []
  [swater]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[Kernels]
  [dummy_pwater]
    type = NullKernel
    variable = pwater
    block = 0
  []
  [dummy_xnacl]
    type = NullKernel
    variable = xnacl
    block = 0
  []
  [dummy_zi]
    type = NullKernel
    variable = zi
    block = 0
  []
[]

[AuxKernels]
  [set_xnacl]
    type = ParsedAux
    variable = aux_xnacl
    expression = 0.35
  []
  [swater]
    type = PorousFlowPropertyAux
    variable = swater
    property = saturation
    phase = 0
    execute_on = timestep_end
  []

[]

[UserObjects]
  [dictator]
    type = PorousFlowDictator
    porous_flow_vars = 'pwater zi xnacl'
    number_fluid_phases = 2
    number_fluid_components = 2
  []
  [pc]
    type = PorousFlowCapillaryPressureConst
    pc = 0 
  []
  [fs]
    type = PorousFlowBrineCO2
    brine_fp = brine
    co2_fp = co2
    capillary_pressure = pc
  []
[]

[FluidProperties]
  [co2]
    type = CO2FluidProperties
  []
  [brine]
    type = BrineFluidProperties
  []
[]

[Materials]
[temperature]
  type = PorousFlowTemperature
  temperature = 20
[]

  [brineco2]
    type = PorousFlowFluidState
    gas_porepressure = pwater
    z = zi
    xnacl = xnacl
    capillary_pressure = pc
    fluid_state = fs
    temperature_unit = Celsius
  []

  [porosity_test]
    type = PorousFlowPorosity_nacl
    block = 0
    fluid = true
    mechanical = false
    thermal = false
    salt_precip = true
    porosity_zero = 0.2 
    reference_porepressure = 1e6
    solid_bulk = 8E9 # unimportant since biot = 1
    xnacl_L = 0.3
    rho_nacl = 2165.0
    r_m = 1.0e-1
    xnacl = aux_xnacl
  []

  [stress_pp_effective]
    type = PorousFlowEffectiveFluidPressure
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  nl_abs_tol = 1e-7
  nl_rel_tol = 1e-5
  l_tol = 1e-5
  l_max_its = 50
  end_time = 100 
  [TimeStepper]
    type = IterationAdaptiveDT
    dt = 0.5
    growth_factor = 1.5
  []
[]

[Outputs]
  exodus = true
  csv = true
[]

[BCs]
  [left]
    type = DirichletBC
    variable = pwater
    boundary = left
    value = 1.2e6
  []
  [right]
    type = DirichletBC
    variable = pwater
    boundary = right
    value = 1e6
  []
  [left_gas]
    type = DirichletBC
    variable = zi
    boundary = left
    value = 0.1
  []
  [right_gas]
    type = DirichletBC
    variable = zi
    boundary = right
    value = 0
  []
[]
```
**3. Further Improvement Suggestions**

Which areas of the current implementation of 'PorousFlowPorosity_nacl' need improvement?

**Comment by robo-boluo** [comment]
I noticed that both Guillaume Giudicelli and Chris Green (@GiudGiud , @cpgr) contributed to the salt precipitation topic in thread #23701. The final collective conclusion was to modify the porosity and permeability file. Therefore, I believe I am heading in the right direction. I look forward to any updates on this topic.

Many thanks.

**Comment by robo-boluo** [comment]
Hi Guillaume Giudicelli @GiudGiud,

I have been working on this problem for quite some time and would like to express my sincere appreciation for your continuous attention and guidance during this period. Initially, I was excited to confirm with you that MOOSE has the capability to model the effect of salt precipitation on porosity and permeability by adjusting the material properties. However, developing the App has proven to be much more complicated. Here is a summary of our progress over the past month:

1. Taking 𝑥NaCl  (NaCl mass fraction) as a variable, I developed the '**PorousFlowPorosity_nacl**' App based on the existing '**PorousFlowPorosity**' App in the MOOSE folder (/moose/modules/porous_flow/). Since the salt precipitation and chemical reaction parts have similar processes affecting porosity, I programmed the salt precipitation by following the format of the chemical part, defining the 'atNegInfinityQp', 'datNegInfinityQp', 'atZeroQp', and 'datZeroQp' sections for my 'PorousFlowPorosity_nacl' app.

2. In the input file, the variable 𝑥NaCl  is computed using type=PorousFlowMassTimeDerivative, type=PorousFlowAdvectiveFlux, and type=PorousFlowDispersiveFlux in [Kernels].

3. I defined 𝑥NaCl_max as the maximum NaCl solubility in the aquifer. Once 𝑥NaCl>𝑥NaCl_max, halite starts to precipitate. To avoid  𝑥NaCl>𝑥NaCl_max, we applied the ConstantBounds to restrict 𝑥NaCl within [0,𝑥NaCl_max].  However, the porosity change was not observed in the simulation results. **"Failed"**

4. We suspected that the ConstantBounds restriction on 𝑥NaCl≤𝑥NaCl_max prevented precipitation from occurring. Therefore, we deleted the ConstantBounds. After precipitation, we should update 𝑥NaCl. Consequently, the ‘XNaclUpdate’ app ([see above](https://github.com/idaholab/moose/discussions/27832#discussioncomment-9809918)) was developed and applied in [Kernels]. However, the expected porosity change was still not achieved. **"Failed"**

5. Even though I increased the 𝑥NaCl removal rate (rate * (𝑥NaCl−𝑥NaCl_max)), it still did not result in any porosity change. We also observed 𝑥NaCl>𝑥NaCl_max and encountered convergence problems at 𝑥NaCl>𝑥NaCl_max. **"Failed"**

Given your experience in MOOSE development, I trust your assertion that MOOSE can handle salt precipitation-induced porosity changes. Therefore, the issue may lie in my implementation of the 'PorousFlowPorosity_nacl' App. I would sincerely appreciate it if you could review my problem and provide your insights ([code link](https://drive.google.com/drive/folders/11RLH6zaP1aGQkqJNdjX2Nctd1dWTgzAb?usp=sharing)).

Thank you for your time and attention to this matter.

**Reply by cpgr** [reply]
Hi @robo-boluo - I'm sorry that I haven't seen this until now. This is something that we have done before! I'm happy to try and help you out with this next week.

**Reply by robo-boluo** [reply]
Dear Chris Green @cpgr 

 I am grateful to your time and generous assistance on this matter. I am looking forward to the further discussion.

**Reply by robo-boluo** [reply]
> Hi @robo-boluo - I'm sorry that I haven't seen this until now. This is something that we have done before! I'm happy to try and help you out with this next week.

Hi Chris Green @cpgr,

May I ask if there are any updates you can share with me at this stage?

Thanks.

**Reply by robo-boluo** [reply]
Dear Chris Green and Guillaume Giudicell @cpgr @GiudGiud ,

Are there any updates on this issue that can be shared at this time? I apologize for the frequent inquiries, but my supervisors are requesting information on the feasibility of this method.
Thanks.

**Reply by GiudGiud** [reply]
I'm not a porous flow module developer
From a general MOOSE developer point of view, I think this should be doable

Remind me of the behavior with the fraction variable? It s supposed to be physically bound? Should it have its own variable?

**Reply by cpgr** [reply]
Hi @robo-boluo, sorry, I haven't made much progress due to other commitments. I'll try and take a look this afternoon and get back to you.

**Reply by cpgr** [reply]
Just as a reminder (like @GiudGiud mentioned), do you want to consider the case where water evaporates into the CO2 gas phase, the salt mass fraction increases until it reaches the solubility limit, then the excess salt precipitates, reducing porosity (from which permeability reduces)?

**Reply by robo-boluo** [reply]
Hi Chris Green @cpgr ,

I appreciate your time and kind help. Yes, this is exactly what I want to implement. There should be mutual solubility between water and CO2. The reduction in porosity and permeability is caused by NaCl precipitation.

**Reply by robo-boluo** [reply]
> I'm not a porous flow module developer From a general MOOSE developer point of view, I think this should be doable
>  Remind me of the behavior with the fraction variable? It s supposed to be physically bound? Should it have its own variable?



The mechanism process of salt precipitation is clear to me. In my code, xnacl represents the mass fraction variable in Kernel block.

**Reply by GiudGiud** [reply]
> the salt mass fraction increases until it reaches the solubility limit

how does this happen in the equations? When the solubility limit is reached, does some sink term sky-rocket or something? Or a property changes to prevent xnacl to keep going?
